### PR TITLE
Octree converter

### PR DESCRIPTION
**Pull request by cphyc**
This adds the ability to convert a list of positions (x, y, z) with levels into an octree that can be loaded by yt.load_octree.

```python
import numpy as np
import yt
from yt_experiments.octree.converter import OctTree
xyz = np.array([
    [0.375, 0.125, 0.125],
    [0.125, 0.125, 0.125],
    [0.375, 0.375, 0.375],
    [0.75, 0.75, 0.75],
])
levels = np.array([2, 2, 2, 1], dtype=np.int32)
data = {
    ("gas", "density"): np.random.rand(4)
}

oct = OctTree.from_list(xyz, levels, check=True)
ref_mask, leaf_order = oct.get_refmask()

for k, v in data.items():
    # Make it 2D so that yt doesn't think those are particles
    data[k] = np.where(leaf_order >= 0, v[leaf_order], np.nan)[:, None]

ds = yt.load_octree(ref_mask, data)
```

### CODE REVIEW

File: yt_experiments/octree/converter.pyx [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+    @classmethod
+    @cython.boundscheck(False)
+    def from_list(cls, const cython.floating[:, ::1] Xc, const cython.integral[::1] levels, bint check = False):
+        """Create an octree from a list of cells.
+
+        Parameters
+        ----------
+        Xc : double array (N, 3)
+            The positions of the cells in unitary units (within [0, 1])
+        levels : int array (N,)
```
Comment (chrishavlin) [inline yt_experiments/octree/converter.pyx:151]: i'm confused by this description (and in the other function docstrings) -- I don't see how `Xc * 2**level` will always be an integer? e.g., with the example:

```python
>>> xyz = np.array(
...             [
...                 [0.375, 0.125, 0.125],
...                 [0.125, 0.125, 0.125],
...                 [0.375, 0.375, 0.375],
...                 [0.75, 0.75, 0.75],
...             ])
>>> levels = np.array([2, 2, 2, 1], dtype=np.int32)
>>> for i in range(0,len(levels)):
...     print(xyz[i,:] * 2**levels[i])
... 
[1.5 0.5 0.5]
[0.5 0.5 0.5]
[1.5 1.5 1.5]
[1.5 1.5 1.5]
```

**Review by chrishavlin — COMMENTED** [review]
Generally looks good to me, but I think there's a memory leak? I haven't been able to pin it down, but here's a little test script:

```python
# mem_leak_test.py
import sys
import numpy as np
from yt_experiments.octree.converter import OctTree

if __name__ == '__main__':
    xyz = np.array(
        [
            [0.375, 0.125, 0.125],
            [0.125, 0.125, 0.125],
            [0.375, 0.375, 0.375],
            [0.75, 0.75, 0.75],
        ],
    )
    levels = np.array([2, 2, 2, 1], dtype=np.int32)

    for _ in range(int(sys.argv[1])):
        octree = OctTree.from_list(xyz, levels, check=True)        
        del octree
```
running with `python mem_leak_test 1000000` (need a large number of iterations since the number of nodes is small), memory usage grows continuously. Ends up using close to 400 Mb before finishing the iterations.

**Comment by chrishavlin** [comment]
Screenshot 2024-07-19 at 3 41 55 PM

e.g., here's a flamegraph from memray for the above example -- the blue and orange lines are resident and heap size over time, y scale on the plot is from 0 to ~500 Mb

**Comment by chrishavlin** [comment]
oh! one more suggestion -- would be good to add `*converter.cpp` to .gitignore

File: yt_experiments/octree/converter.pyx (hunk 1 above)
Comment (cphyc) [reply yt_experiments/octree/converter.pyx:163]: My mistake, it should be half-integers. I've fixed the docstring now :)

**Comment by cphyc** [comment]
There was indeed a memory leak (I didn't implement a deallocate step). I've just checked with your code snippet, and the memory use is now constant. Thanks for the thorough review!

**Comment by chrishavlin** [comment]
Great! Ya, I thought you needed a `__deallocate__` but i wasn't sure about whether it needed to recursively walk through the children (which based on your implementation now I see that it does!).

**Review by chrishavlin — APPROVED** [review]

**Pull request merged**